client.go

```diff
@@ -2134,6 +2195,16 @@ func (c *Client) writePublicationUpdatePosition(ch string, pub *protocol.Publica
 	currentPositionOffset := channelContext.streamPosition.Offset
 	nextExpectedOffset := currentPositionOffset + 1
 	pubOffset := pub.Offset
+	pubEpoch := sp.Epoch
+	if pubEpoch != channelContext.streamPosition.Epoch {
+		if c.node.logger.enabled(LogLevelDebug) {
+			c.node.logger.log(newLogEntry(LogLevelDebug, "client insufficient state", map[string]interface{}{"channel": ch, "user": c.user, "client": c.uid, "epoch": pubEpoch, "expectedEpoch": channelContext.streamPosition.Epoch}))
+		}
+		// Oops: sth lost, let client reconnect to recover its state.
+		go func() { _ = c.close(DisconnectInsufficientState) }()
```

Comment:
is it ok that `close` happens not under the lock?

Reply:
yeah, there are locks inside (at least I believe that it's properly synchronized) and it's ok that actual close will happen somewhere in future. Closing is a bit tricky in client to make sure we get an expected order of callbacks (make sure we do not have alive event callback after disconnect event callback for example).